sycl/source/exception.cpp

```diff
@@ -10,9 +10,105 @@
+                             // error code)
+std::error_code sycl121_proxy_errorcode = make_error_code(sycl::errc::invalid);
+} // namespace
+
+exception::exception(std::error_code ec, const char *Msg)
+    : exception(static_cast<context *>(nullptr), ec, Msg) {}
+
+exception::exception(std::error_code ec, const std::string &Msg)
+    : exception(static_cast<context *>(nullptr), ec, Msg) {}
+
+// new SYCL2020 constructors
+exception::exception(std::error_code ec)
+    : exception(static_cast<context *>(nullptr), ec, "") {}
+
+exception::exception(int ev, const std::error_category &ecat,
+                     const std::string &what_arg)
+    : exception(static_cast<context *>(nullptr), {ev, ecat}, what_arg) {}
+
+exception::exception(int ev, const std::error_category &ecat,
+                     const char *what_arg)
+    : exception(static_cast<context *>(nullptr), {ev, ecat},
+                std::string(what_arg)) {}
+
+exception::exception(int ev, const std::error_category &ecat)
+    : exception(static_cast<context *>(nullptr), {ev, ecat}, "") {}
+
+exception::exception(context ctx, std::error_code ec,
+                     const std::string &what_arg)
+    : exception(&ctx, ec, what_arg) {}
+
+exception::exception(context ctx, std::error_code ec, const char *what_arg)
+    : exception(ctx, ec, std::string(what_arg)) {}
+
+exception::exception(context ctx, std::error_code ec)
+    : exception(ctx, ec, "") {}
+
+exception::exception(context ctx, int ev, const std::error_category &ecat,
+                     const char *what_arg)
+    : exception(ctx, {ev, ecat}, std::string(what_arg)) {}
+
+exception::exception(context ctx, int ev, const std::error_category &ecat,
+                     const std::string &what_arg)
+    : exception(ctx, {ev, ecat}, what_arg) {}
+
+exception::exception(context ctx, int ev, const std::error_category &ecat)
+    : exception(ctx, ev, ecat, "") {}
+
+// protected base constructor for all SYCL2020 constructors
+exception::exception(context *ctxPtr, std::error_code ec,
+                     const std::string &what_arg)
+    : MMsg(what_arg + reserved_for_errorcode), MCLErr(PI_INVALID_VALUE),
+      MContext(ctxPtr) {
+  // For compatibility with previous implementation, we are "hiding" the
+  // std:::error_code in the MMsg string, behind the null string terminator
+  size_t whatLen = what_arg.length();
+  char *reservedPtr = &MMsg[whatLen];
+  reservedPtr[0] = '\0';
+  reservedPtr++;
+  // insert 'SYCL' marker.
+  reservedPtr[0] = 'S';
+  reservedPtr[1] = 'Y';
+  reservedPtr[2] = 'C';
+  reservedPtr[3] = 'L';
+  reservedPtr += 4;
+  // insert error code
+  std::error_code *ecPtr = reinterpret_cast<std::error_code *>(reservedPtr);
+  *ecPtr = ec;
+}
+
+const std::error_code &exception::code() const noexcept {
+  const char *whatStr = MMsg.c_str();
+  size_t whatLen = strlen(whatStr);
+  // advance past null string-terminator
+  const char *reservedPtr = &whatStr[whatLen + 1];
```

Comment:
I don't think this works as you expect. Technically `what_arg` is std::string, which can have null terminator in the middle of the string, so strlen here won't give you what you expect.

Reply:
Thanks! Good catch. I reworked that.